pmd-core/src/main/java/net/sourceforge/pmd/RuleSet.java

```diff
@@ -122,15 +122,24 @@ public RuleSet(final RuleSet rs) {
          * Add a new rule to this ruleset. Note that this method does not check
          * for duplicates.
          *
-         * @param rule
+         * @param newRule
          *            the rule to be added
          * @return The same builder, for a fluid programming interface
          */
-        public RuleSetBuilder addRule(final Rule rule) {
-            if (rule == null) {
+        public RuleSetBuilder addRule(final Rule newRule) {
+            if (newRule == null) {
                 throw new IllegalArgumentException(MISSING_RULE);
             }
-            rules.add(rule);
+
+            // check for duplicates - adding more than one rule with the same name will
+            // be problematic - see #RuleSet.getRuleByName(String)
+            for (Rule rule : rules) {
+                if (rule.getName().equals(newRule.getName()) && rule.getLanguage() == newRule.getLanguage()) {
+                    LOG.warning("Duplicated rule: " + newRule.getName());
```

Comment:
we should at least `break` here, possible even `return this` and avoid the duplicate, shouldn't we?

Reply:
Hm... I'm not sure, if it's worth it. I added this check and log, as I was curious whether this would ever happen. I think, we should for 7.0.0 prevent to load such rulesets, since these rulesets are broken IMHO.

Current behaviour is (since rules is a List) both rules are executed. It could be the same rule (as the name indicates), but it could also be a completely different rule with the same name (or same rule with different properties).
Returning here would change the result.

But yes, breaking here makes sense - the first duplicate is enough.

@jsotuyod  Maybe we should phrase the warning a bit more drastic?

`The rule with name XYZ is duplicated. Future versions of PMD will reject to load such rulesets.`